File: scripts/weather.lua
function kola_main(city)
	local url = "http://weather.hao.360.cn/api_weather_info.php?app=hao360"

	if city == '' then
		c = kola.getarea()


		if c then
			city = string.format("%s-%s-%s", c.province, c.city, c.city)
		end
	end

	if city ~= nil and city ~= '' then
		url = kola.geturl("/city?cid=1&name=" .. city)
		local areaCode = kola.wget(url)

		if areaCode ~= nil and areaCode ~= '' then
			url = string.format("http://weather.hao.360.cn/api_weather_info.php?app=hao360&city_code=%s", areaCode)
		end
	end
	local text = kola.wget(url, false)
	--print(text)

	local ret = {}
	if text then
		text = kola.pcre("callback\\((.*)\\);", text)
		local js = cjson.decode(text)

		if js == nil then
			return {}
		end

		ret.area = {}
		city = ''
		k=1
		for _,v in pairs(js.area) do
			if v[1] ~= city then
				ret.area[k] = v[1]
				city = v[1]
				k = k + 1
			end
		end

		ret.weather = {}
		local idx = 1
		for i, w in pairs(js.weather) do
			info = {}

			if w.info ~= "" then
				info.date = w.date

				info.day = {}
				info.day.picture       = string.format("http://p2.qhimg.com/d/_hao360/weather/big/%s.png",  w.info.day[1])
				info.day.code          = w.info.day[1]
				info.day.weather       = w.info.day[2]
				info.day.temp          = w.info.day[3]
				info.day.windDirection = w.info.day[4]
				info.day.windPower     = w.info.day[5]

				info.night = {}
				info.night.picture       = string.format("http://p2.qhimg.com/d/_hao360/weather/big/%s.png",  w.info.night[1])
				info.night.code          = w.info.night[1]
				info.night.weather       = w.info.night[2]
				info.night.temp          = w.info.night[3]
				info.night.windDirection = w.info.night[4]
				info.night.windPower     = w.info.night[5]
				ret.weather[idx] = info
				idx = idx + 1
			end
		end
		if js.pm25 and js.pm25.pm25 then
			ret.pm25 = tostring(js.pm25.pm25[1])
		end
	end

	--print(cjson.encode(ret))
	return cjson.encode(ret)
end

File: sdk/weather.cpp
#include "kola.hpp"
#include "http.hpp"
#include "json.hpp"
#include "script.hpp"

KolaWeather::~KolaWeather()
{
	Clear();
}

void KolaWeather::Clear()
{
	size_t count = weatherList.size();
	
	for (int i=0; i < count; i++)
		delete weatherList[i];
	
	weatherList.clear();
}

void KolaWeather::Get(string name, StringList &value)
{
	string url = "city?name=" + name;

	json_t *js = json_loadurl(url.c_str());

	if (js) {
		json_t *v;
		json_array_foreach(js, v) {
			const char *n = json_gets(v, "name", "");
			value.Add(n);
		}

		json_decref(js);
	}
}

void KolaWeather::GetProvince(StringList &value)
{
	Get("", value);
}

void KolaWeather::GetCity(string province, StringList &city)
{
	Get(province, city);
}

void KolaWeather::GetCounty(string province, string city, StringList &county)
{
	Get(province + "-" + city, county);
}

void KolaWeather::SetArea(string area)
{
	this->Area = area;
}

void KolaWeather::SetArea(string province, string area, string county)
{
	if (not province.empty()) {
		this->Area = province;

		if (not area.empty()) {
			this->Area += "-" + area;

			if (not county.empty())
				this->Area += "-" + county;
		}
	}
}

bool KolaWeather::ParserWeatherData(WeatherData& data, json_t *js)
{
	data.picture       = json_gets(js, "picture",       "");
	data.code          = json_gets(js, "code",          "");
	data.weather       = json_gets(js, "weather",       "");
	data.temp          = json_gets(js, "temp",          "");
	data.windDirection = json_gets(js, "windDirection", "");
	data.windPower     = json_gets(js, "windPower",     "");

	return true;
}

void KolaWeather::Run(void)
{
	LuaScript &lua = LuaScript::Instance();
	vector<string> args;

	args.push_back(Area);
	string text = lua.RunScript(args, "weather");

	if (not text.empty()) {
		json_error_t error;

		json_t *js = json_loads(text.c_str(), JSON_DECODE_ANY, &error);
		
		if (js == NULL)
			return;

		mutex.lock();
		Clear();

		json_t *weather = json_geto(js, "weather");
		if (weather) {
			json_t *info;
			json_array_foreach(weather, info) {
				Weather *w = new Weather();
				
				w->date = json_gets(info, "date", "");
				
				json_t *v = json_geto(info, "day");
				if (v)
					ParserWeatherData(w->day, v);
				v = json_geto(info, "night");
				if (v)
					ParserWeatherData(w->night, v);

				this->weatherList.push_back(w);
			}
		}

		PM25 = json_gets(js, "pm25", "");
		mutex.unlock();

		json_delete(js);
	}
}

bool KolaWeather::UpdateFinish()
{
	return status == Task::StatusFinish;
}

void KolaWeather::Update()
{
	if (status != Task::StatusDownloading) {
		mutex.lock();
		Clear();
		mutex.unlock();
		status = Task::StatusInit;
		Start();
	}
}

Weather *KolaWeather::Today()
{
	Weather *w = NULL;
	mutex.lock();
	if (weatherList.size() > 0)
		w = weatherList[0];
	mutex.unlock();
	return w;
}

Weather *KolaWeather::Tomorrow()
{
	Weather *w = NULL;
	mutex.lock();
	if (weatherList.size() > 1)
		w = weatherList[1];
	
	mutex.unlock();
	return w;
}

